File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.py
class Solution:
    def myAtoi(self, s: str) -> int:
        i = 0
        n = len(s)
        sign = 1
        result = 0

        # removing white spaces
        while i < n and s[i] ==' ':
            i += 1
        # assinging sign
        if i < n and (s[i] == '+' or s[i] == '-'):
            if s[i] == '-':
                sign = -1
            i += 1
        # Calculating numbers
        while i < n and s[i].isdigit():
            digit = int(s[i])

            # Handling overflow
            if result > (2**31 - 1 - digit) // 10:
                return 2**31-1 if sign == 1 else -2**31
            result = result * 10 + digit
            i += 1
        return sign * result

File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.java
class Solution {
    public int myAtoi(String s) {
        int i = 0;
        int n = s.length();
        int sign = 1;
        int result = 0;

        // removing whitespace
        while (i < n && s.charAt(i) == ' ') {
            i++;
        }
        // checkimh the sign
        if(i < n && (s.charAt(i) == '+' || s.charAt(i) == '-')) {
            if(s.charAt(i) == '-') sign = -1;
            i++; 
        }
        // caculating the number
        while (i < n && Character.isDigit(s.charAt(i))) {
            int digit = s.charAt(i) - '0';
            // handling the overflow
            if (result > (Integer.MAX_VALUE - digit) / 10) {
                return sign == 1 ? Integer.MAX_VALUE : Integer.MIN_VALUE;
            }
            result = result * 10 + digit;
            i++;
        }
        return sign * result;

    }
}